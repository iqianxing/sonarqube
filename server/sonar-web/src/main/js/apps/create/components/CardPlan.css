/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.card-plan {
  display: flex;
  flex-direction: column;
  width: 450px;
  height: 210px;
  background-color: #fff;
  border: solid 1px var(--barBorderColor);
  border-radius: 3px;
  box-sizing: border-box;
  margin-right: calc(2 * var(--gridSize));
  transition: all 0.2s ease;
}

.card-plan.animated {
  height: 0;
  border-width: 0;
  overflow: hidden;
}

.card-plan.animated.open {
  height: 210px;
  border-width: 1px;
}

.card-plan.highlight {
  box-shadow: var(--defaultShadow);
}

.card-plan:last-child {
  margin-right: 0;
}

.card-plan:focus {
  outline: none;
}

.card-plan-actionable {
  cursor: pointer;
}

.card-plan-actionable:not(.disabled):hover {
  box-shadow: var(--defaultShadow);
  transform: translateY(-2px);
}

.card-plan-actionable.selected {
  border-color: var(--darkBlue);
}

.card-plan-actionable.selected .card-plan-recommended {
  border: solid 1px var(--darkBlue);
  border-top: none;
}

.card-plan-actionable.disabled {
  cursor: not-allowed;
  background-color: var(--disableGrayBg);
  border-color: var(--disableGrayBorder);
}

.card-plan-actionable.disabled h2,
.card-plan-actionable.disabled ul {
  color: var(--disableGrayText);
}

.card-plan-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: calc(2 * var(--gridSize)) calc(2 * var(--gridSize)) 0;
}

.card-plan-price {
  font-size: var(--bigFontSize);
}

.card-plan-body {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 0 calc(2 * var(--gridSize)) calc(2 * var(--gridSize));
}

.card-plan-body .alert {
  margin-bottom: 0;
}

.card-plan-recommended {
  position: relative;
  padding: 6px calc(var(--gridSize) * 2);
  left: -1px;
  bottom: -1px;
  width: 450px;
  color: #fff;
  background-color: var(--blue);
  border-radius: 0 0 3px 3px;
  box-sizing: border-box;
  font-size: var(--smallFontSize);
}
