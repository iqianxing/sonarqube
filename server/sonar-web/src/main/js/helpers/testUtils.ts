/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { InjectedRouter } from 'react-router';
import { Location } from 'history';
import { EditionKey } from '../apps/marketplace/utils';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', { ...mockEvent, ...event });
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.body.clientWidth/clientHeight` are getters by default, so we need to redefine them
  // pass `configurable: true` to allow to redefine the properties multiple times
  if (width) {
    Object.defineProperty(document.body, 'clientWidth', { configurable: true, value: width });
  }
  if (height) {
    Object.defineProperty(document.body, 'clientHeight', { configurable: true, value: height });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}

export function mockLocation(overrides = {}): Location {
  return {
    action: 'PUSH',
    key: 'key',
    pathname: '/path',
    query: {},
    search: '',
    state: {},
    ...overrides
  };
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}

export function mockAppState(overrides = {}): T.AppState {
  return {
    defaultOrganization: 'foo',
    edition: EditionKey.community,
    productionDatabase: true,
    qualifiers: ['TRK'],
    settings: {},
    version: '1.0',
    ...overrides
  };
}

export function mockComponent(overrides = {}): T.Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    organization: 'foo',
    qualifier: 'TRK',
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way'
      }
    ],
    tags: [],
    ...overrides
  };
}

export function mockCurrentUser(overrides = {}): T.CurrentUser {
  return {
    isLoggedIn: true,
    ...overrides
  };
}

export function mockOrganization(overrides = {}): T.Organization {
  return {
    key: 'foo',
    name: 'Foo',
    ...overrides
  };
}
