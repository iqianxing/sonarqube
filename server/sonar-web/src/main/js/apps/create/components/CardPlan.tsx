/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import UpgradeOrganizationAdvantages from './UpgradeOrganizationAdvantages';
import RecommendedIcon from '../../../components/icons-components/RecommendedIcon';
import { Alert } from '../../../components/ui/Alert';
import { formatPrice } from '../organization/utils';
import { translate } from '../../../helpers/l10n';
import './CardPlan.css';

interface Props {
  className?: string;
  disabled?: boolean;
  onClick?: () => void;
  selected?: boolean;
  startingPrice?: number;
}

interface CardProps extends Props {
  children: React.ReactNode;
  recommended?: string;
  title: string;
}

export default function CardPlan(props: CardProps) {
  const { className, disabled, onClick, recommended, selected, startingPrice } = props;
  const isActionable = Boolean(onClick);
  return (
    <div
      aria-checked={selected}
      className={classNames(
        'card-plan',
        { 'card-plan-actionable': isActionable, disabled, selected },
        className
      )}
      onClick={isActionable && !disabled ? onClick : undefined}
      role="radio"
      tabIndex={0}>
      <h2 className="card-plan-header big-spacer-bottom">
        <span className="display-flex-center">
          {isActionable && (
            <i className={classNames('icon-radio', 'spacer-right', { 'is-checked': selected })} />
          )}
          {props.title}
        </span>
        {startingPrice !== undefined &&
          (startingPrice ? (
            <FormattedMessage
              defaultMessage={translate('billing.price_from_x')}
              id="billing.price_from_x"
              values={{
                price: <span className="card-plan-price">{formatPrice(startingPrice)}</span>
              }}
            />
          ) : (
            <span className="card-plan-price">{formatPrice(0)}</span>
          ))}
      </h2>
      <div className="card-plan-body">{props.children}</div>
      {recommended && (
        <div className="card-plan-recommended">
          <RecommendedIcon className="spacer-right" />
          <FormattedMessage
            defaultMessage={recommended}
            id={recommended}
            values={{ recommended: <strong>{translate('recommended')}</strong> }}
          />
        </div>
      )}
    </div>
  );
}

interface FreeProps extends Props {
  almName?: string;
  hasWarning: boolean;
}

export function FreeCardPlan({ almName, hasWarning, ...props }: FreeProps) {
  const showInfo = almName && props.disabled;
  const showWarning = almName && hasWarning && !props.disabled;

  return (
    <CardPlan startingPrice={0} title={translate('billing.free_plan.title')} {...props}>
      <>
        <div className="spacer-left">
          <ul className="big-spacer-left note">
            <li className="little-spacer-bottom">
              {translate('billing.free_plan.all_projects_analyzed_public')}
            </li>
            <li>{translate('billing.free_plan.anyone_can_browse_source_code')}</li>
          </ul>
        </div>
        {showWarning && (
          <Alert variant="warning">
            <FormattedMessage
              defaultMessage={translate('billing.free_plan.private_repo_warning')}
              id="billing.free_plan.private_repo_warning"
              values={{ alm: almName }}
            />
          </Alert>
        )}
        {showInfo && (
          <Alert variant="info">
            <FormattedMessage
              defaultMessage={translate('billing.free_plan.not_available_info')}
              id="billing.free_plan.not_available_info"
              values={{ alm: almName }}
            />
          </Alert>
        )}
      </>
    </CardPlan>
  );
}

interface PaidProps extends Props {
  isRecommended: boolean;
}

export function PaidCardPlan({ isRecommended, ...props }: PaidProps) {
  return (
    <CardPlan
      recommended={isRecommended ? translate('billing.paid_plan.recommended') : undefined}
      title={translate('billing.paid_plan.title')}
      {...props}>
      <>
        <UpgradeOrganizationAdvantages />
        <div className="big-spacer-left">
          <Link className="spacer-left" target="_blank" to="/about/pricing">
            {translate('billing.pricing.learn_more')}
          </Link>
        </div>
      </>
    </CardPlan>
  );
}
