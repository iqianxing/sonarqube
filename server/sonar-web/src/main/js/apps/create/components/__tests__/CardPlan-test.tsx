/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import CardPlan, { FreeCardPlan, PaidCardPlan } from '../CardPlan';
import { click } from '../../../../helpers/testUtils';

it('should render correctly', () => {
  expect(
    shallow(
      <CardPlan recommended="Recommended for you" startingPrice={10} title="Paid Plan">
        <div>content</div>
      </CardPlan>
    )
  ).toMatchSnapshot();
});

it('should be actionable', () => {
  const onClick = jest.fn();
  const wrapper = shallow(
    <CardPlan onClick={onClick} title="Free Plan">
      <div>content</div>
    </CardPlan>
  );

  expect(wrapper).toMatchSnapshot();
  click(wrapper);
  wrapper.setProps({ selected: true, startingPrice: 0 });
  expect(wrapper).toMatchSnapshot();
});

describe('#FreeCardPlan', () => {
  it('should render', () => {
    expect(shallow(<FreeCardPlan hasWarning={false} />)).toMatchSnapshot();
  });

  it('should render with warning', () => {
    expect(
      shallow(<FreeCardPlan almName="GitHub" hasWarning={true} selected={true} />)
    ).toMatchSnapshot();
  });

  it('should render disabled with info', () => {
    expect(
      shallow(<FreeCardPlan almName="GitHub" disabled={true} hasWarning={false} />)
    ).toMatchSnapshot();
  });
});

describe('#PaidCardPlan', () => {
  it('should render', () => {
    expect(shallow(<PaidCardPlan isRecommended={true} startingPrice={10} />)).toMatchSnapshot();
  });
});
