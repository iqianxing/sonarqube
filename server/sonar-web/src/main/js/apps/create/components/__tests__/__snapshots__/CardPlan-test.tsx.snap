// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#FreeCardPlan should render 1`] = `
<CardPlan
  startingPrice={0}
  title="billing.free_plan.title"
>
  <div
    className="spacer-left"
  >
    <ul
      className="big-spacer-left note"
    >
      <li
        className="little-spacer-bottom"
      >
        billing.free_plan.all_projects_analyzed_public
      </li>
      <li>
        billing.free_plan.anyone_can_browse_source_code
      </li>
    </ul>
  </div>
</CardPlan>
`;

exports[`#FreeCardPlan should render disabled with info 1`] = `
<CardPlan
  disabled={true}
  startingPrice={0}
  title="billing.free_plan.title"
>
  <div
    className="spacer-left"
  >
    <ul
      className="big-spacer-left note"
    >
      <li
        className="little-spacer-bottom"
      >
        billing.free_plan.all_projects_analyzed_public
      </li>
      <li>
        billing.free_plan.anyone_can_browse_source_code
      </li>
    </ul>
  </div>
  <Alert
    variant="info"
  >
    <FormattedMessage
      defaultMessage="billing.free_plan.not_available_info"
      id="billing.free_plan.not_available_info"
      values={
        Object {
          "alm": "GitHub",
        }
      }
    />
  </Alert>
</CardPlan>
`;

exports[`#FreeCardPlan should render with warning 1`] = `
<CardPlan
  selected={true}
  startingPrice={0}
  title="billing.free_plan.title"
>
  <div
    className="spacer-left"
  >
    <ul
      className="big-spacer-left note"
    >
      <li
        className="little-spacer-bottom"
      >
        billing.free_plan.all_projects_analyzed_public
      </li>
      <li>
        billing.free_plan.anyone_can_browse_source_code
      </li>
    </ul>
  </div>
  <Alert
    variant="warning"
  >
    <FormattedMessage
      defaultMessage="billing.free_plan.private_repo_warning"
      id="billing.free_plan.private_repo_warning"
      values={
        Object {
          "alm": "GitHub",
        }
      }
    />
  </Alert>
</CardPlan>
`;

exports[`#PaidCardPlan should render 1`] = `
<CardPlan
  recommended="billing.paid_plan.recommended"
  startingPrice={10}
  title="billing.paid_plan.title"
>
  <UpgradeOrganizationAdvantages />
  <div
    className="big-spacer-left"
  >
    <Link
      className="spacer-left"
      onlyActiveOnIndex={false}
      style={Object {}}
      target="_blank"
      to="/about/pricing"
    >
      billing.pricing.learn_more
    </Link>
  </div>
</CardPlan>
`;

exports[`should be actionable 1`] = `
<div
  className="card-plan card-plan-actionable"
  onClick={[MockFunction]}
  role="radio"
  tabIndex={0}
>
  <h2
    className="card-plan-header big-spacer-bottom"
  >
    <span
      className="display-flex-center"
    >
      <i
        className="icon-radio spacer-right"
      />
      Free Plan
    </span>
  </h2>
  <div
    className="card-plan-body"
  >
    <div>
      content
    </div>
  </div>
</div>
`;

exports[`should be actionable 2`] = `
<div
  aria-checked={true}
  className="card-plan card-plan-actionable selected"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "currentTarget": Object {
              "blur": [Function],
            },
            "preventDefault": [Function],
            "stopPropagation": [Function],
            "target": Object {
              "blur": [Function],
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  role="radio"
  tabIndex={0}
>
  <h2
    className="card-plan-header big-spacer-bottom"
  >
    <span
      className="display-flex-center"
    >
      <i
        className="icon-radio spacer-right is-checked"
      />
      Free Plan
    </span>
    <span
      className="card-plan-price"
    >
      billing.price_format.0
    </span>
  </h2>
  <div
    className="card-plan-body"
  >
    <div>
      content
    </div>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
<div
  className="card-plan"
  role="radio"
  tabIndex={0}
>
  <h2
    className="card-plan-header big-spacer-bottom"
  >
    <span
      className="display-flex-center"
    >
      Paid Plan
    </span>
    <FormattedMessage
      defaultMessage="billing.price_from_x"
      id="billing.price_from_x"
      values={
        Object {
          "price": <span
            className="card-plan-price"
          >
            billing.price_format.10
          </span>,
        }
      }
    />
  </h2>
  <div
    className="card-plan-body"
  >
    <div>
      content
    </div>
  </div>
  <div
    className="card-plan-recommended"
  >
    <RecommendedIcon
      className="spacer-right"
    />
    <FormattedMessage
      defaultMessage="Recommended for you"
      id="Recommended for you"
      values={
        Object {
          "recommended": <strong>
            recommended
          </strong>,
        }
      }
    />
  </div>
</div>
`;
